### ISSUE TITLE
2017.7.0 master upgrade breaks mine data on non-glob matching on minions

### ISSUE THREAD

**Issue by boltronics** [saltstack/salt#42713]
### Description of Issue/Question
Note: Somebody might want to throw up a big warning at the top of the release notes before other people upgrading get bitten by this.

I've been rolling out the new 2017.7.0 to staging, which despite some obvious problems (such as orchestration being broken due to being unable to read Pillar data, https://github.com/saltstack/salt/issues/42403), it had been going okay - although being unable to use orchestrate makes the deployment process painfully slow.

Our environment in AWS includes a HAProxy load balancer, and a number of backend app servers running Apache. When an app is ready to receive requests, it adds itself to the Mine, which in turn makes itself known to HAProxy by way of config file modification and a service reload.

As per the [release nodes](https://docs.saltstack.com/en/develop/topics/releases/2017.7.0.html#remember-to-update-the-salt-master-first), we started the migration to the new Salt release (from 2016.11.4 in our case) to 2017.7.0 by upgrading the master first. From the docs "Backwards compatiblity[sic] is still maintained, so older minions can still be used". However this does not appear to be completely true.

During staging, we tested a rapid roll-out of the new Salt version by bringing up new instances with the new version pre-installed. In production, we took a longer time to roll everything out. Perhaps this is why we didn't encounter this issue on staging, if mine data on the minion was cached and hadn't refreshed... not sure. But in production, we had a HAProxy instance (running 2016.11.4 against a 2017.7.0 master) get its configuration file updated, but couldn't see any mine data so omitted all the server entries. This left us with a configuration file that was unable to proxy traffic. Our site was brought down.

I quickly picked up the issue (running `sudo salt '*' mine.get ...` on the master) and observed that all instances not yet upgraded were experiencing this issue.

Perhaps now would be a good time to bring up https://github.com/saltstack/salt/issues/14080 again? I appreciate everyone's work, but having to explain Salt broke our infrastructure again is frustrating.

### Setup
Install a 2016.11.4 master running two or more 2016.11.4 minions, and have all of them put data in the mine (eg. an IP address). Now upgrade the master and one minion to 2017.7.0. a `mine.update` might be required... but somehow the 2016.11.4 minion will at some point stop being able to see any mine data. As I have since discovered (just while writing this report in fact), this only occurs if not using glob matching. Glob matching (regardless of 'glob' being explicitly specified or not) continues to work. Unfortunately we use pcre matching extensively.

### Steps to Reproduce Issue
If I run `sudo salt '*' mine.get '*' some.module` I get all expected results from all hosts. If I run `sudo salt '*' mine.get '.*' some.module pcre`, I only receive results from updated minions. Likewise, I can run `sudo salt '*' mine.get 'osrelease:8.9' some.module grain` and get the same problematic results.

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

Upgraded 2017.7.0 master:
```
Salt Version:
           Salt: 2017.7.0
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.3.7
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.9 
         locale: UTF-8
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.9 
 ```

Upgraded 2017.7.0 minion:
```
Salt Version:
           Salt: 2017.7.0
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.3.7
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.9 
         locale: UTF-8
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.9 
 ```

Minion not yet upgraded and affected by the upgrade:
```
Salt Version:
           Salt: 2016.11.4
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.2
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.9 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.9 

```

All hosts were running Debian Jessie.

**Comment by damon-atkins** [comment]
Salt Team did provide release candidates. It would be interesting see what the download count is for 2017.7.0 vs 2017.7.0 release candidates to see how much community testing occurred.

**Comment by boltronics** [comment]
Indeed.

**Comment by garethgreenaway** [comment]
@boltronics Thanks for the report.  To clarify, you've upgraded your master to 2017.7.0 but one minion remains at 2016.11.x and is experiencing problems when running mine.get?

**Comment by boltronics** [comment]
@garethgreenaway Our salt-master was upgraded, correct. But this was an oversimplified example. In practice, while most of our instances had been upgraded, we had maybe 5 or 6 still on 2016.11.4 since we wanted to be extra careful with those. This meant we had a CI server that could no longer deploy since it no longer knew the addresses of the instances to deploy to, for example. But obviously bringing our website down was the worst, so that's why I just mentioned the one managing our HAProxy config - to highlight the seriousness of this issue.

We quickly finished off the roll-out to avoid further problems, so all are on 2017.7.0 as of now.

**Comment by garethgreenaway** [comment]
Understood.  We do our best to ensure that everything works and to maintain backward compatibility between releases, unfortunately, in the case of running different versions for master and minions there are a lot of things that likely won't work as expected. This release did have more package testing but only from the perspective of the minions & master being the same version.  It's possible that we'll get a point where we can test additional cases such as having minions of the previous release.

**Comment by boltronics** [comment]
Fair enough. I was under the impression it was a supported setup for upgrading purposes, although I knew it wouldn't have been tested to the same extent. Thanks for clarifying the position of the project. Hopefully one day the upgrade roll-out process can be tested in an automated fashion so it's not such a burden.

**Comment by gracinet** [comment]
Hi there, I've rediscovered this issue independently in my testing lab (minions with 2016.11, master with 2017.7). I think (evidence below) that it's caused by the change from `expr_form` to `tgt_type`, and that it could be easily fixed.

Simply put, no other selecting method (called ``expr_form`` in Carbon and ``tgt_type`` in Nitrogen) than the default glob can work.

Even though the release notes clearly state that for Nitrogen, the whole fleet has to be upgraded at once, it can't always be done in practice, and certainly not atomically.  I think the Mine is important enough that a 

Here's an excerpt of ``salt/modules/mine.py`` (version 2016.11.8):

```python
    load = {
            'cmd': '_mine_get',
            'id': __opts__['id'],
            'tgt': tgt,
            'fun': fun,
            'expr_form': expr_form,
    }
    ret = _mine_get(load, __opts__)
```

On a 2016.11 minion, I tried changing the `expr_form` key to `tgt_type` in that `load` dict, and it started working right away.
I also tried to *add* `tgt-type`, giving this 

```python
     load = {
            'cmd': '_mine_get',
            'id': __opts__['id'],
            'tgt': tgt,
            'fun': fun,
            'expr_form': expr_form,
            'tgt_type': expr_form,
    }
    ret = _mine_get(load, __opts__)
```
and it also worked.

Obviously this requires further testing, and I did not look closely at the master-side Mine code. It's quite possible that such simple a change could be made, and therefore a fix done in the Nitrogen series only (compared to all currently supported versions on the minion side)
